### ISSUE TITLE
Champion: `base(T)` phase two

### ISSUE THREAD

**Issue by gafter** [dotnet/csharplang#2337]
- [X] Proposal added (below)
- [X] Discussed in LDM https://github.com/dotnet/csharplang/blob/master/meetings/2019/LDM-2019-02-27.md
- [ ] Decision in LDM
- [ ] Finalized (done, rejected, inactive)
- [ ] Spec'ed

On 2019-02-27 the LDM met to make some decisions about the *default interface methods* feature.

[Plan A]: We decided that the best design for the `base(I).M` feature would be that we lookup `M` in the interface type `I`, and the result must be accessible.  Moreover, there is a requirement that if the found member is abstract, the type `I` must have a unique most specific (concrete) implementation in the type `I`.  The compiler would emit IL that identifies the method found, and the type `I`, and the runtime would select the most specific (concrete) implementation in the type `I` and invoke it (or throw an exception if there is no unique most specific implementation at runtime).  There is no IL defined today that would serve this purpose, so we would have to design it.  (One option would be to generate a "constrained" prefix with the interface type `I`).  The IL should not assume or require that the implementing method is accessible to the caller (e.g. it could be private), but it does require that the declared method named in the IL is accessible.

Unfortunately, we do not believe we have the resources to design and implement the feature in this form  (plan A) before we'd like to deliver its first implementation (hopefully in C# 8.0).  So initially, we'll ship a language feature that is more constrained.  This is our plan B for C# 8.0:

[Plan B]: We lookup `M` in the interface type `I`, and the result must be accessible.  Moreover, there is a requirement that the found member must have an implementation directly in the type `I`, and that implementation method must be accessible.  The compiler will produce IL that directly (non-virtually) invokes that method.

We expect to ship plan B in C# 8.0.

This issue is for moving to plan A in a subsequent language version.

**Comment by YairHalberstadt** [comment]
Would it be possible to add an example of what would be possible under plan A, but is not under plan B? Thanks

**Comment by Joe4evr** [comment]
If I understand correctly:
```cs
public interface IFoo
{
    int M();
}
public interface IBar : IFoo
{
    int IFoo.M() => 42;
}
public interface IBaz : IFoo
{
    int X() => base(IFoo).M(); //requires Plan A
}
```

**Comment by YairHalberstadt** [comment]
@Joe4Evr
Was that meant to be `IBaz : IBar`?

**Comment by Joe4evr** [comment]
Not necessarily. Under Plan A, the method implementation could come from somewhere else and would be looked up at runtime. But I'll adjust anyway.

**Comment by YairHalberstadt** [comment]
In that case what's the difference between calling base(IFoo).M and call M directly? A case when IBar or IBaz hides M?

**Comment by gafter** [comment]
@Joe4evr Not quite
``` c#
public interface I1
{
    int M();
}
public interface I2 : I1
{
    int I1.M() => 42;
}
public interface I3 : I2
{
}
public interface C : I3
{
    int X1() => base(I2).M(); // Plan B
    int X2() => base(I3).M(); // requires Plan A
}
```

**Comment by Thaina** [comment]
Sorry that it might be too late but could we change syntax to `base<T>` instead of `base(T)`

I still feel awkward to have type in parentheses and thinking that syntax related with type should always be `<T>` if possible

**Comment by HaloFour** [comment]
@Thaina 

https://github.com/dotnet/csharplang/blob/master/meetings/2018/LDM-2018-11-14.md

**Comment by Thaina** [comment]
@HaloFour Are there anywhere I could look for reason behind this decision?

**Comment by Joe4evr** [comment]
https://github.com/dotnet/csharplang/blob/master/meetings/2017/LDM-2017-06-14.md#syntax

**Comment by Thaina** [comment]
@Joe4evr Thank you very much

**Comment by gafter** [comment]
There is a thread starting at https://github.com/dotnet/csharplang/issues/406#issuecomment-495905388 with some users noting how they would really like to use this functionality.  I'm posting this here in case it is useful in prioritizing our candidates for C# 9.0.

**Comment by gafter** [comment]
There is a draft specification for `base(T)` at https://github.com/dotnet/csharplang/issues/2910

**Comment by IllidanS4** [comment]
This seems to be especially useful for structs, as calling an explicit interface implementation via a base lookup would remove the need for boxing to the speciifc interface type, and thus also allow modifications of the original value done by the called method.

**Comment by Thaina** [comment]
@gafter So you mean DIM will always box struct?

**Comment by gafter** [comment]
Yes.  We tried to design it so it would not, but it would have made things much more complicated.

**Comment by Thaina** [comment]
Thank you, well, yet another reason I should try to avoid DIM as much as possible

**Comment by gafter** [comment]
@Thaina how do you imagine this boxing would occur without the base(T) feature?

**Comment by Thaina** [comment]
@gafter Or did you mean normal DIM will not box struct, only when we did override DIM and then try to call the original interface DIM?

**Comment by IllidanS4** [comment]
@gafter I was referring to this case:
```cs
struct Struct : IEnumerable<object>
{
    public IEnumerator<object> GetEnumerator()
    {
        yield break;
    }

    IEnumerator IEnumerable.GetEnumerator()
    {
        yield break;
    }

    public IEnumerator GetNonGenericEnumerator()
    {
        return ((IEnumerable)this).GetEnumerator(); //boxes
        return base(IEnumerable).GetEnumerator(); //calls IEnumerable.GetEnumerator directly
    }
}
```
Seems like that's what should happen, as `base(T)`, as I understand it, should call the method on `this` but perform the lookup on a base class or interface.

Of course one could argue that this code should be designed the other way around, but I've also ran into this situation with partial structs where the auto-generated part expected the struct to implement a particular set of generic interfaces, differing only in the return type for one method, I couldn't simply create a new custom-named method for every one of them.

**Comment by gafter** [comment]
You appear to completely misunderstand what a base call does.  It calls the *implementation* in the named type.

-Neal
________________________________

**Comment by IllidanS4** [comment]
My mistake, I just forgot that interfaces can contain implementation. But still, if the interface doesn't provide the implementation, this construct could be used to refer to the implementation present on the current type, couldn't it?

**Comment by gafter** [comment]
No

**Comment by JinShil** [comment]
The [default interface members proposal](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-8.0/default-interface-methods#base-interface-invocations) had an example that allowed for overriding default implementations:

```C#
using static System.Console;

interface IA
{
    void M() 
    {
        WriteLine("IA.M"); 
    }
}
interface IB : IA
{
    override void IA.M()    // The modifier 'override' is not valid for this item
    {
        WriteLine("IB.M");
    }
}
interface IC : IA
{
    override void IA.M()    // The modifier 'override' is not valid for this item
    { 
        WriteLine("IC.M"); 
    }
}
```
Currently that results in compiler errors as shown above.  Did the ability to override implementaions get cut from C#8 along with `base(T)` or are the two on different implementation paths?

**Comment by Joe4evr** [comment]
@JinShil See the first post:
> Update: [LDM 04-29-2019](https://github.com/dotnet/csharplang/blob/master/meetings/2019/LDM-2019-04-29.md#default-interface-implementations-and-base-calls) cut `base()` from C# 8.0.

**Comment by gafter** [comment]
@JinShil Drop the `override` keyword.  I think that should work.

**Comment by JinShil** [comment]
Just a friendly ping on this issue.  Looking at the list of upcoming C# 11 features, they aren't nearly as valuable, and even somewhat superficial, compared to the ability to call base default interface methods.  Default interface methods will always be incomplete without this feature, and those creating complex hierarchies, and wishing to maximize code reuse, will always run into this limitation.  Due to this limitation I have begun using source generators as a substitute.  Please finish this implementation.

**Comment by zvrba** [comment]
Any progress on this? I stumbled upon another issue that'd be solved by base invocation: https://github.com/dotnet/csharplang/discussions/8269

**Comment by westy** [comment]
Agree that this is badly needed, since could be a massive source of bugs in years to come when interfaces are used without detailed knowledge of their implementation!

/subscribed
